### ISSUE TITLE
SshExec fails despite exit code 0 if stderr contains output

### ISSUE THREAD

**Issue by kriegaex** [mojohaus/wagon-maven-plugin#4]
This fails just because echo writes something to stderr:

``` xml
              <execution>
                <id>exec-ssh-script</id>
                <phase>install</phase>
                <goals>
                  <goal>sshexec</goal>
                </goals>
                <configuration>
                  <url>${scp.url}</url>
                  <commands>
                    <command>echo "Test 1"</command>
                    <command>echo "Test 2" &#62;&#38;2</command>
                  </commands>
                  <displayCommandOutputs>true</displayCommandOutputs>
                </configuration>
              </execution>
```

Expected behaviour: On exit code 0 no exception should be raised, no matter whether there is log output on stderr or not.

I do not know the code base and have no idea how to implement Maven plugins or mojos, but is this somehow related to [this code snippet](https://github.com/apache/maven-wagon/blob/f3462c3f276b2449689eda1f52c3fa1f8e2177ae/wagon-providers/wagon-ssh/src/main/java/org/apache/maven/wagon/providers/ssh/jsch/AbstractJschWagon.java#L333) from Maven Wagon?

``` java
if ( streams.getErr().length() > 0 && !ignoreFailures )
{
    int exitCode = channel.getExitStatus();
    throw new CommandExecutionException( "Exit code: " + exitCode + " - " + streams.getErr() );
}
```

I did not find anything like this in the mojo itself, but it has got to come from somewhere.

**Comment by dantran** [comment]
very likely, maven-wagon not able to detect  remote exit code, it relies heavily on stderr .  I think you may end  up to set the ignore option.

**Comment by kriegaex** [comment]
Well, if it is unable to detect the remote exit code, then it should be changed to have this capability. I actually see no reason why it should not be possible to do so. An SSH connection is just a remote shell and you should have all means to determine exit codes you also have in a local shell session.

**Comment by dantran** [comment]
does throw exception when exit code is not 0? I have an impression wagon never able to detect non zero exit code

**Comment by kriegaex** [comment]
I tested the behaviour like this:

```
<command>ls xxyy 2>&amp;1</command>
```

Even though the exit code is 2, no error is detected by Wagon SSH just because I redirected standard error to standard output. But your assumption is still incorrect because Wagon *is* able to detect exit codes. It just does not handle them when nothing is logged on standard error. Proof:

```
<command>ls xxyy 2>&amp;1</command>
```

This yields the following output:

```
[ERROR] Failed to execute goal org.codehaus.mojo:wagon-maven-plugin:1.0:sshexec (exec-ssh-script) on project sdv.okv.aggregate: Unable to execute remote command: Exit code: 2 - ls: cannot access xxyy: No such file or directory -> [Help 1]
```

See? "Exit code: 2". Wagon gets the exit code, it just handles it wrong. So please improve the plugin, would you? Thank you.

**Comment by dantran** [comment]
filed https://issues.apache.org/jira/browse/WAGON-465

**Comment by dantran** [comment]
Unfortunately even the  I fix upstream wagon-ssh, this plugin cant pickup the fix since it is bound to the runtime maven. 

in compromise, i can fix this plugin to honor 'failOnError' base on stderr

**Comment by kriegaex** [comment]
Admittedly I do not fully understand your last comment and also not what you did in WAGON-465 because there are no links I can follow to review the source code changes (but it is not my job anyway, it would just have helped me understand what you have done). From what I see it seems that you added the ability to ignore exit codes in addition to stdErr output. But the problem I describe is rather the opposite: exit codes are already ignored even though they should not be. I want stdErr to be ignored, but exit codes honoured because the latter are the relevant thing.

Maybe you did what I requested and I just misinterpret your statements. So please forgive me if I am just saying stupid things, but please explain in more detail so as to make me understand. :-)

**Comment by dantran** [comment]
wagon-maven-plugin (unfortunately) is bound to the maven-wagon embedded under the maven runtime that runs this plugin.  So my fix at maven-wagon, which introduce one more 'execute' interface with option to  ignore stdErr and/or exit code, will not get picked up by this plugin until next version of maven perhaps

so if you really want the option to fail on exit code,  you will need to create your own plugin to interface directly with wagon-ssh 2.11-SNAPSHOT using my newly 'execute' interface ( just borrow wagon:sshexec code

**Comment by kriegaex** [comment]
Nah, I guess we will not do that. But as soon as there is a downloadable Maven snapshot using the fixed Wagon version, feel free to let us know so we can provide test feedback. :-)

**Comment by dantran** [comment]
fixed at 4b2c331,   please test and your maven env must proxy both sonatype and apache snapshot repository

**Comment by kriegaex** [comment]
After our admin gavwe me access to the snaphot repos I have retested and everything looks good.

Specifically, these two succeed:

    <command>echo "standard out"</command>
    <command>echo "standard error" &#62;&#38;2</command>

And these two fail:

    <command>ls non-existing-file</command>
    <command>ls non-existing-file &#62;&#38;1</command>

Looking forward to the new release. :-)

**Comment by ykcycer** [comment]
Sorry for the lame question: how to pass new parameters?

**Comment by kriegaex** [comment]
Sorry for the lame answer: I do not understand your question. And how it is connected to this issue?

**Comment by ykcycer** [comment]
You got me. The question is really unclear.
I meant WAGON-465 introduced two new parameters ignoreStderr and ignoreNoneZeroExitCode. How to configure pom.xml to pass them? I wanna skip std output errors, but check exit codes in my project.

PS: use a coffee machine

**Comment by tchemit** [comment]
@dantran Hi, could we use the ```CommandExecutor``` instead of ```ScpWagon```, so should extends some modification to have same behaviour you had on ScpWagon. Can you manage this ? (this will solved #12) thanks
